File: dsa-practice-ram/dsa-practice-ram/Arrays/LongestConsecutive.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading.Tasks;

namespace dsa_practice_ram.Arrays
{
    internal class LongestConsecutive
    {
        public void LongestConsecutiveSequenceSlidingWindow()
        {
            int[] nums = { 102, 4, 4, 4, 4, 4, 4, 44, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 100, 1, 101, 3, 2, 1, 1 };
            // sorted: 1, 1, 1, 2, 3, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 44, 100, 101, 102
            // Time complexity: nLog(n)

            Array.Sort(nums);

            int maxLength = 1;
            Range range = default;

            int windowStart = 0;

            for (int windowEnd = 1; windowEnd < nums.Length; windowEnd++)
            {
                int curr = nums[windowEnd];
                int previous = nums[windowEnd - 1];

                if (curr - previous != 1)
                {
                    windowStart = windowEnd;
                }

                int windowLength = windowEnd - windowStart + 1;
                if (windowLength > maxLength)
                {
                    maxLength = windowLength;
                    range = windowStart..(windowEnd + 1);
                }
            }

            Console.WriteLine(string.Join(", ", nums[range]));
        }
        
        public void LongestConsecutiveSequenceHashsetApproach()
        {
            int[] nums = { 102, 4, 4, 4, 4, 4, 4, 44, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 100, 1, 101, 3, 2, 1, 1 };
            // sorted: 1, 1, 1, 2, 3, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 44, 100, 101, 102
            // Time complexity: O(n)

            HashSet<int> seen = new HashSet<int>(nums);

            int maxLength = 0;

            foreach (var num in nums)
            {
                if (!seen.Contains(num - 1))
                {
                    int current = num;
                    int count = 0;

                    while (seen.Contains(current))
                    {
                        current++;
                        count++;
                    }
                    maxLength = Math.Max(maxLength, count);
                }
            }

            Console.WriteLine(maxLength);
        }
    }
}

File: MyLearning/PythonPractice/dsa_practice/sliding_window/longest_consecutive.py
"""Longest consecutive sequence utilities."""

# C# LongestConsecutive.LongestConsecutiveSequenceSlidingWindow

def longest_consecutive_sequence_sliding_window(nums=None):
    """Return the longest consecutive sequence after sorting.

    Problem: Find the longest sequence of consecutive integers.
    Approach: Sort then track a sliding window of consecutive values.
    Time complexity: O(n log n)
    Space complexity: O(1) extra (in-place sort)
    """
    if nums is None:
        nums = [102, 4, 4, 4, 4, 4, 4, 44, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 100, 1, 101, 3, 2, 1, 1]

    if not nums:
        return []

    nums = list(nums)
    nums.sort()

    max_length = 1
    best_range = (0, 1)
    window_start = 0

    for window_end in range(1, len(nums)):
        curr = nums[window_end]
        previous = nums[window_end - 1]

        if curr - previous != 1:
            window_start = window_end

        window_length = window_end - window_start + 1
        if window_length > max_length:
            max_length = window_length
            best_range = (window_start, window_end + 1)

    return nums[best_range[0]:best_range[1]]


# C# LongestConsecutive.LongestConsecutiveSequenceHashsetApproach

def longest_consecutive_sequence_hashset(nums=None):
    """Return the length of the longest consecutive sequence using a set.

    Problem: Find the longest consecutive integer run.
    Approach: Use a set to expand sequences from their starting numbers.
    Time complexity: O(n)
    Space complexity: O(n)
    """
    if nums is None:
        nums = [102, 4, 4, 4, 4, 4, 4, 44, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 100, 1, 101, 3, 2, 1, 1]

    if not nums:
        return 0

    seen = set(nums)
    max_length = 0

    for num in nums:
        if num - 1 not in seen:
            current = num
            count = 0

            while current in seen:
                current += 1
                count += 1

            if count > max_length:
                max_length = count

    return max_length
